### PR TITLE
Add Fortran version of requires dynamic_allocators test

### PR DESCRIPTION
**Pull request by jhdavis8**
Currently will not compile with any available compiler on Summit, there is some difficulty in detecting issues with the test.

### CODE REVIEW

File: sys/systems/summit.def [hunk 1]
```diff
@@ -29,6 +29,7 @@ endif
 # Clang compiler
 # ifeq ($(CC), clang)
 #   C_COMPILER_MODULE = llvm/1.0-20190225
+#   C_COMPILER_MODULE = cuda/10.1.243; module use /sw/summit/modulefiles/ums/stf010/Core; module load llvm/12.0.0-latest
```
Comment (spophale) [inline sys/systems/summit.def:32]: Change to llvm/13.0.0-latest

**Comment by spophale** [comment]
What are the compiler errors/warnings ?

**Comment by jhdavis8** [comment]
gfortran error:
```
/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90:26:9:                                                                                      [30/1965]

   26 |   !$omp requires dynamic_allocators
      |         1
Error: Unclassifiable OpenMP directive at (1)
/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90:39:36:

   39 |     INTEGER(omp_memspace_handle_kind):: x_memspace = omp_default_mem_space
      |                                    1
Error: Symbol 'omp_memspace_handle_kind' at (1) has no IMPLICIT type
/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90:40:25:

   40 |     type(omp_alloctrait):: x_traits(1) = &[omp_alloctrait(omp_atk_alignment,64)]
      |                         1
Error: Derived type 'omp_alloctrait' at (1) is being used before it is defined
/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90:41:37:

   41 |     INTEGER(omp_allocator_handle_kind):: x_alloc
      |                                     1
Error: Symbol 'omp_allocator_handle_kind' at (1) has no IMPLICIT type
/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90:43:18:

   43 |     !$omp target defaultmap(tofrom)
      |                  1
Error: Failed to match clause at (1)
/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90:46:11:

   46 |     !$omp allocate(x) allocator(x_alloc)
      |           1
Error: Unclassifiable OpenMP directive at (1)
/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90:47:15:

   47 |     allocate(x(N))
      |               1
Error: Shape specification for allocatable scalar at (1)
/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90:60:20:

   60 |     !$omp end target
      |                    1
Error: Unexpected !$OMP END TARGET statement at (1)
/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90:62:17:

   62 |     deallocate(x)
      |                 1
Error: 'x' at (1) is not a variable
/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90:44:11:

   44 |     x_alloc = omp_init_allocator(x_memspace, 1, x_traits)
      |           1
Error: Symbol 'x_alloc' at (1) has no IMPLICIT type
/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90:44:43:

   44 |     x_alloc = omp_init_allocator(x_memspace, 1, x_traits)
      |                                           1
Error: Symbol 'x_memspace' at (1) has no IMPLICIT type
/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90:44:56:

   44 |     x_alloc = omp_init_allocator(x_memspace, 1, x_traits)
      |                                                        1
Error: Symbol 'x_traits' at (1) has no IMPLICIT type
/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90:50:38:

   50 |     !$omp do simd simdlen(16) aligned(x: 64)
      |                                      1
Error: Object 'x' is not a variable at (1)
/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90:57:47:

   57 |        OMPVV_TEST_AND_SET_VERBOSE(errors, x(i) .ne. i)
      |                                               1
Error: The allocatable object 'x' at (1) must have an explicit function interface or be declared as array
```

XLF error:
```
** ompvv_lib   === End of Compilation 1 ===
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90", line 26.9: 1515-019 (S) Syntax is incorrect.
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90", line 39.13: 1516-050 (S) Expression or initial value must be evaluated at compile time.
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90", line 39.13: 1516-036 (S) Entity omp_memspace_handle_kind has undefined type.
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90", line 39.54: 1516-050 (S) Expression or initial value must be evaluated at compile time.
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90", line 39.54: 1516-036 (S) Entity omp_default_mem_space has undefined type.
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90", line 40.42: 1515-019 (S) Syntax is incorrect.
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90", line 40.42: 1515-018 (S) Incorrect character found in source at line 40 in column 42.  Hexadecimal value of character is 26.
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90", line 41.13: 1516-036 (S) Entity omp_allocator_handle_kind has undefined type.
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90", line 41.13: 1516-050 (S) Expression or initial value must be evaluated at compile time.
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90", line 43.18: 1515-022 (S) Syntax Error: Extra token " defaultmap " was found. The token is ignored.
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90", line 43.28: 1515-022 (S) Syntax Error: Extra token " ( " was found. The token is ignored.
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90", line 43.29: 1515-022 (S) Syntax Error: Extra token " tofrom " was found. The token is ignored.
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90", line 43.35: 1515-022 (S) Syntax Error: Extra token " ) " was found. The token is ignored.
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90", line 44.15: 1516-036 (S) Entity omp_init_allocator has undefined type.
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90", line 46.10: 1515-019 (S) Syntax is incorrect.
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90", line 47.14: 1516-026 (S) Non-arrays must not be subscripted.
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90", line 52.8: 1516-026 (S) Non-arrays must not be subscripted.
"/autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/requires/test_requires_dynamic_allocators.F90", line 57.47: 1516-026 (S) Non-arrays must not be subscripted.
** test_requires_dynamic_allocators   === End of Compilation 2 ===
1501-511  Compilation failed for file test_requires_dynamic_allocators.F90.
```

File: sys/systems/summit.def (hunk 1 above)
Comment (jhdavis8) [reply sys/systems/summit.def]: Changed it to llvm 13 and added some required flags.

**Comment by tob2** [comment]
May I propose the following change:
```diff
--- test_requires_dynamic_allocators.F90.orig   2021-03-04 23:36:44.606587866 +0100
+++ test_requires_dynamic_allocators.F90        2021-03-04 23:38:28.726429865 +0100
@@ -35,8 +35,8 @@
   INTEGER FUNCTION test_dynamic_allocators()
     INTEGER:: errors = 0
-    INTEGER,ALLOCATABLE:: x
+    INTEGER,ALLOCATABLE:: x(:)
     INTEGER:: i
     INTEGER(omp_memspace_handle_kind):: x_memspace = omp_default_mem_space
-    type(omp_alloctrait):: x_traits(1) = &[omp_alloctrait(omp_atk_alignment,64)]
+    type(omp_alloctrait):: x_traits(1) = omp_alloctrait(omp_atk_alignment,64)
     INTEGER(omp_allocator_handle_kind):: x_alloc
```

Additionally, code like:
```Fortran
INTEGER(omp_memspace_handle_kind):: x_memspace = omp_default_mem_space
```
means that 'x_memspace' has implicitly the SAVE attribute, i.e. it is in static memory. I think that's okay, but I just want to mention it.

Side note: With GCC 11's gfortran and the fix above, it has fewer fails, but the testcase still does not fully compile. ("Sorry, ‘dynamic_allocators’ clause at (1) on REQUIRES directive is not yet supported" and 'defaultmap' and 'omp allocate' fail.)

**Comment by jhdavis8** [comment]
Thanks for the suggestions @tob2, they should be added.

I notice that the [OpenMP examples](https://www.openmp.org/wp-content/uploads/openmp-examples-5-0-1.pdf) on the bottom half of page 377 use the `&[ ]` syntax removed by your second fix, I wonder why that is?

**Comment by tob2** [comment]
> I notice that the [OpenMP 5.0 examples](https://www.openmp.org/wp-content/uploads/openmp-examples-5-0-1.pdf) on the bottom half of page 377 use the `&[ ]` syntax removed by your second fix, I wonder why that is?

That OpenMP example is fine:
* The `&` indicates that the line is continued in the next line – but that only works if after the ampersand there is only whitespace or a comment starts ("!"). — Thus, in your example only a line break was missing. (Sorry for not spotting it.)
* And `[...]` was actually fine. The left-hand side is a one-element array and the RHS was / is in the OpenMP example a one-element array constructor. By removing the `[...]`, we now have a scalar. And `array = scalar` works, assigning the same value to all elements, of which there is one.

Maybe it is a bit cleaner to re-add the `[...]` (but array(1)=scalar also works). If you want you can also re-add the `&` but only if followed by a line break.

**Review by tmh97 — APPROVED** [review]

**Review by spophale — APPROVED** [review]

**Pull request merged**